### PR TITLE
Impose requirements on descriptors from spirv image types

### LINKED ISSUE #762: Validation suggestion: textureSamples() should not be called for single-sampled image bindings
We've just discovered a problem with one of the recently introduced CTS tests, where they first bind a single-sample MS image and then try to query the number of samples exposed by the image with a textureSamples() call. This would work in GLSL, but is illegal after being converted to SPIR-V due to the following restriction in the SPIR-V spec:

```
OpImageQuerySamples
Query the number of samples available per texel fetch in a multisample image.
Result Type must be a scalar integer type. The result is the number of samples.
Image must be an object whose type is OpTypeImage. Its Dim operand must be one
of 2D and MS of 1(multisampled)
```

SPIRV-Validation does not have shadow enough state information to determine this condition, so validation layers seem like a more suitable candidate.

### PR DESCRIPTION
**Pull request by chrisforbes**
Fixes #762, and lays a lot of groundwork for other image usage validation that @tobine has wanted for a while.

### CODE REVIEW

File: layers/descriptor_sets.cpp [hunk 1]
```diff
@@ -327,18 +327,31 @@ cvdescriptorset::DescriptorSet::~DescriptorSet() {
...
 }
+
+
+static char const * string_descriptor_req_view_type(descriptor_req req) {
+    for (unsigned i = 0; i <= VK_IMAGE_VIEW_TYPE_END_RANGE; i++) {
+        if (req & (1 << i)) {
+            return string_VkImageViewType(VkImageViewType(i));
+        }
+    }
+}
```
Comment (tobine) [inline layers/descriptor_sets.cpp]: error: control may reach end of non-void function

Comment (lentinem) [inline layers/descriptor_sets.cpp]: Can we auto upload build failures?

**Comment by lentinem** [comment]
+1

**Comment by chrisforbes** [comment]
Updated and pushed to master.

**Pull request merged**